src/eduid_webapp/reset_password/views/reset_password.py

```diff
@@ -0,0 +1,369 @@
+    except BadCode as e:
+        return error_message(e.msg)
+
+    salt = session.reset_password.generated_password_salt
+    hashed = b64encode(hash_password(password, salt)).decode('utf8')
+    if hashed == session.reset_password.generated_password_hash:
+        state.generated_password = True
+        current_app.logger.info('Generated password used')
+        current_app.stats.count(name='reset_password_generated_password_used')
+    else:
+        state.generated_password = False
+        current_app.logger.info('Custom password used')
+        current_app.stats.count(name='reset_password_custom_password_used')
+
+    current_app.logger.info(f'Resetting password for user {state.eppn}')
+    reset_user_password(state, password)
+    current_app.logger.info(f'Password reset done, removing state for user {state.eppn}')
+    current_app.password_reset_state_db.remove_state(state)
+    return success_message('resetpw.pw-resetted')
+
+
+@reset_password_views.route('/extra-security/', methods=['POST'])
+@UnmarshalWith(ResetPasswordExtraSecSchema)
+@MarshalWith(FluxStandardAction)
+def choose_extra_security(code: str, phone_index: int) -> dict:
+    """
+    View called when the user chooses extra security (she can do that when she
+    has some verified phone number). It receives an emailed reset password code
+    and an index for one of the verified phone numbers, and returns info on the
+    result of the attempted operation.
+
+    Preconditions required for the call to succeed:
+    * A PasswordResetEmailState object in the password_reset_state_db
+      keyed by the received code.
+    * A flag in said state object indicating that the emailed code has already
+      been verifed.
+    * The user referenced in the state has at least phone_index (number) of
+      verified phone numbers.
+
+    As side effects, this operation will:
+    * Copy the data in the PasswordResetEmailState to a new
+      PasswordResetEmailAndPhoneState;
+    * Create a new random hash as identifier code for the new state;
+    * Store this code in the new state;
+    * Send an SMS message with the code to the phone number corresponding to
+      the received phone_index;
+
+    This operation may fail due to:
+    * The code does not correspond to a valid state in the db;
+    * The code has expired;
+    * No valid user corresponds to the eppn stored in the state;
+    * Problems sending the SMS message
+    """
+    try:
+        state = get_pwreset_state(code)
+    except BadCode as e:
+        return error_message(e.msg)
+
+    current_app.logger.info(f'Password reset: choose_extra_security for {state.eppn}')
+
+    # Check that the email code has been validated
+    if not state.email_code.is_verified:
+        current_app.logger.info(f'User {state.eppn} has not verified their email address')
+        return error_message('resetpw.email-not-validated')
+
+    phone_number = state.extra_security['phone_numbers'][phone_index]
```

Comment:
1 - Ugh - indexes. 
2 - Shouldn't an SMS be sent to all the users verified phones, in case someone has access to one of them but not another?

Actually, shouldn't _all_ the users verified phones _and_ verified e-mails be notified when extra security is used?

Reply:
At this point, we are sending the extra security code in an SMS to a phone number that the user has chosen, out of all the verified phone numbers she had; So what we do here atm is the right thing.